wpilibj/src/main/java/edu/wpi/first/wpilibj/GenericHID.java

```diff
@@ -159,7 +166,8 @@ public boolean getRawButtonReleased(int button) {
    * @return an event instance representing the button's digital signal attached to the given loop.
    */
   public BooleanEvent button(int button, EventLoop loop) {
-    return new BooleanEvent(loop, () -> getRawButton(button));
+    var cache = m_buttonCache.computeIfAbsent(loop, k -> new HashMap<>());
+    return cache.computeIfAbsent(button, k -> new BooleanEvent(loop, () -> getRawButton(k)));
```

Comment:
While we can assume CommandJoystick functions are called in a single thread, we probably can't assume the same for Joystick and GenericHID functions.  As such, we should mutex-protect these functions (for Java, just marking them synchronized is probably easiest, although the Concurrent version of HashMap might work too).  C++ will need a wpi::mutex with scoped_lock.

Reply:
Can you elaborate on the reason for the difference?
What's the risk here if we don't mutex?

Reply:
Any reason we can't assume Joystick and GenericHID functions will be called in a single thread? Within the context of robot programs, I doubt the few teams that don't use commands will be accessing HID state from multiple threads.

Reply:
Command-based robots are explicitly single-threaded and all command methods are expected to be called from that single thread, so it's a natural extension from that to the Command HID classes.

The main reason these could be called on a separate thread is if a user is running multiple EventLoops.  All other HID functions are thread safe, and these are also currently thread safe, so why break that when it's easy enough to mutex protect them?

Reply:
How can running multiple EventLoops cause these functions to be called from a separate thread? It's not like users are putting EventLoops on separate threads.

Reply:
Is there a reason users can't put EventLoops onto separate threads?

Reply:
Per [Discord](https://discord.com/channels/176186766946992128/368993897495527424/1267014099787714642), I've added mutexes for GenericHID in Java, and no mutex is needed for C++ to prevent multithreaded bugs that are nearly impossible to debug.